src/util/cache.cpp

```diff
@@ -12,7 +12,32 @@ void assertion_cache::run_on_assert(symex_target_equationt::SSA_stept &step)
     step.cond = constant_bool2tc(trivial_value);
   }
   else
+  {
     log_debug("Cache missed");
+    if(match_subexpressions && step.cond->expr_id == expr2t::implies_id)
+    {
+      /* Assertions are in the form of step.guard => (exec_guard => expr)
+       * meaning that we can cache the 'expr' as long as we don't touch 'exec_guard'
+       */
+      auto logic_op = std::dynamic_pointer_cast<logic_2ops>(step.cond);
+      // First, add current expr into DB
+      db.emplace(std::make_pair(logic_op->side_2, step.guard));
+      // Now, check whether any of expr operands
+      if(try_matching_sub_expression(logic_op->side_2, step.guard))
```

Comment:
I assume `logic_op->side_2` refers to `(exec_guard => expr)` in the comment above, yes? If so, how does passing this implication to the algorithm performing the substitutions ensure
> that we can cache the 'expr' as long as we **don't touch 'exec_guard'**

as per the comment above (emphasis mine)? try_matching_sub_expression() could touch exec_guard, can it not?